### PR TITLE
Check if port is occupied before starting Kolibri

### PR DESCRIPTION
**Pull request by lyw07**
### Summary

This PR is to implement one of the sanity checks before starting Kolibri.
In this PR, before launching Kolibri:
1. I check if a Kolibri server has been running using `server.get_status()`. If a Kolibri server is running, then users should not be able to run `kolibri start` again. The program will exit with an error message.
2. If there is no Kolibri server running currently, then I continue to check if the port is occupied, since it could be possible that something other than Kolibri occupies the port. If the port is occupied, the program will exit with an error message.

Additionally, I added a line in the function `stop()` of `cli.py` to indicate that kolibri server has been stopped.

### Reviewer guidance

1. Start a kolibri instance using installers other than the ones from this PR
2. Get an installer from this PR and try to start another kolibri instance

Running with two pex files could be an easier approach. 

### References

Issues: 
https://github.com/learningequality/kolibri/issues/3285
https://github.com/learningequality/kolibri/issues/2867
https://github.com/learningequality/kolibri/issues/3048

----

### Contributor Checklist

- [X] PR has the correct target branch and milestone
- [X] PR has 'needs review' or 'work-in-progress' label
- [X] Contributor has fully tested the PR manually
- [ ] Screenshots of any front-end changes are in the PR description
- [X] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')

### Reviewer Checklist

- [ ] Automated test coverage is satisfactory
- [ ] Reviewer has fully tested the PR manually
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependencies files were updated (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Link to diff of internal dependency change is included
- [ ] CHANGELOG.rst is updated for high-level changes
- [ ] Contributor is in AUTHORS.rst

### CODE REVIEW

File: kolibri/utils/cli.py [hunk 1]
```diff
@@ -314,6 +312,7 @@ def stop():
         pid, __, __ = server.get_status()
         server.stop(pid=pid)
         stopped = True
+        logger.info("Kolibri server has been successfully stoppped.")
```
Comment (indirectlylit) [inline kolibri/utils/cli.py]: > Kolibri server has successfully been stopped.

File: kolibri/utils/sanity_checks.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+    try:
+        s.bind((host, port))
+        s.close()
+    except socket.error:
+        # Port is occupied
+        logger.error(
+            "Port {} is occupied.\n"
+            "Please check that you do not have other processes "
+            "running on this port and try again.\n".format(port))
```
Comment (aronasorman) [inline kolibri/utils/sanity_checks.py]: Note that the recommended indentation style here, is to put the closing parens in its own line. It's to allow people to add in more arguments easily. 

So it'll look something like this:

```
logger.error(
             "Port {} is occupied.\n"
             "Please check that you do not have other processes "
             "running on this port and try again.\n".format(port)
)
```

File: kolibri/utils/sanity_checks.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+logger = logging.getLogger(__name__)
+
+def check_other_kolibri_running(port):
+    try:
+        # Check if there are other kolibri instances running
+        # If there are, then we need to stop users from starting kolibri again.
+        pid, listen_address, listen_port = get_status()
+        logger.error(
+            "There is another Kolibri server running. "
+            "Please use `kolibri stop` and try again.")
```
Comment (aronasorman) [inline kolibri/utils/sanity_checks.py]: Put the parens on their own line her too.

**Comment by indirectlylit** [comment]
looks great!

image

File: kolibri/utils/sanity_checks.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+def check_other_kolibri_running(port):
+    try:
+        # Check if there are other kolibri instances running
+        # If there are, then we need to stop users from starting kolibri again.
+        pid, listen_address, listen_port = get_status()
+        logger.error(
+            "There is another Kolibri server running. "
+            "Please use `kolibri stop` and try again."
+        )
+        sys.exit(1)
```
Comment (benjaoming) [inline kolibri/utils/sanity_checks.py:19]: Just flagging: This is where we need to start defining the actual return codes such that UIs can easily find out why Kolibri doesn't start.

Comment (lyw07) [reply kolibri/utils/sanity_checks.py:19]: Thanks for the information! Maybe raising the exception instead of `sys.exit(1)` would be good?

Comment (benjaoming) [reply kolibri/utils/sanity_checks.py:19]: As you show in this great PR, the startup checks are specific to exactly what happens. Therefor, I would think that raising exceptions and catching them in a systematic way to make program execution terminate will be hard. The user guiding error message is very important :)

Comment (benjaoming) [reply kolibri/utils/sanity_checks.py:19]: What I mean is that here there is a port availability check, which is important to `start`. Maybe also `restart`. But it's not necessary for `manage`.

Comment (lyw07) [reply kolibri/utils/sanity_checks.py:19]: Thank you! I just realized what you mean about the return codes. As for the `restart`, I'm sorry that I didn't notice this before. But it seems that `restart` is not really restarting the server right now?

**Pull request merged**